torch/testing/_internal/common_utils.py

```diff
@@ -69,6 +69,113 @@
 IS_FBCODE = os.getenv('PYTORCH_TEST_FBCODE') == '1'
 IS_REMOTE_GPU = os.getenv('PYTORCH_TEST_REMOTE_GPU') == '1'
 
+
+class _TestParametrizer(object):
+    """
+    Decorator class for parametrizing a test function, yielding a set of new tests spawned
+    from the original generic test, each specialized for a specific set of test inputs. For
+    example, parametrizing a test across the set of ops will result in a test function per op.
+
+    The decision of how to parametrize / what to parametrize over is intended to be implemented
+    by each derived class.
+
+    In the details, the decorator adds a 'parametrize_fn' property to the test function that is called
+    during device-specific test instantiation performed in instantiate_device_type_tests(). Because of this,
+    there is no need to parametrize over device type, as that is already handled separately.
+    """
+    def _parametrize_test(self, test, generic_cls, device_cls):
+        """
+        Parametrizes the given test function across whatever dimension is specified by the derived class.
+        Tests can be parametrized over any arbitrary dimension or combination of dimensions, such as all
+        ops, all modules, or all ops + their associated dtypes.
+
+        Args:
+            test (fn): Test function to parametrize over
+            generic_cls (class): Generic test class object containing tests (e.g. TestFoo)
+            device_cls (class): Device-specialized test class object (e.g. TestFooCPU); set to None
+                if the tests are not part of a device-specific set
+
+        Returns:
+            Generator object returning 3-tuples of:
+                test (fn): Parametrized test function; must support a device arg and args for any params
+                test_name (str): Parametrized name of the test (e.g. test_bar_opname_int64)
+                param_kwargs (dict): Param kwargs to pass to the test (e.g. {'op': 'add', 'dtype': torch.int64})
+        """
+        raise NotImplementedError
+
+    def __call__(self, fn):
+        fn.parametrize_fn = self._parametrize_test
```

Comment:
Would this allow for multiple @parametrize to test the product of some parameters?

Reply:
Currently, no. I'm looking into two types of composability for `@parametrize`:
1. Product-wise composability as you mentioned (test product of multiple args)
2. Additive composability (each `@parametrize` instance specifies all args, adding to the tests that are run)

Technically, it might be possible to have both, but I'm leaning towards only providing the latter. I think if a true product is desired, it can be generated inline manually. Do you have a strong opinion either way?

Reply:
Update here: I ended up adding product-wise composability since several people I talked to expected it.